core/src/main/java/org/elasticsearch/index/query/PercolatorQueryBuilder.java

```diff
@@ -0,0 +1,379 @@
+        if (indexedDocumentVersion != null) {
+            getRequest.version(indexedDocumentVersion);
+        }
+        GetResponse getResponse = queryShardContext.getClient().get(getRequest).actionGet();
+        if (getResponse.isExists() == false) {
+            throw new ResourceNotFoundException(
+                    "indexed document [{}/{}/{}] couldn't be found", indexedDocumentIndex, indexedDocumentType, indexedDocumentId
+            );
+        }
+        return new PercolatorQueryBuilder(documentType, getResponse.getSourceAsBytesRef());
+    }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        if (indexedDocumentIndex != null || indexedDocumentType != null || indexedDocumentId != null) {
+            throw new IllegalStateException("query builder must be rewritten first");
+        }
+
+        if (document == null) {
+            throw new IllegalStateException("nothing to percolator");
+        }
+
+        MapperService mapperService = context.getMapperService();
+        DocumentMapperForType docMapperForType = mapperService.documentMapperWithAutoCreate(documentType);
+        DocumentMapper docMapper = docMapperForType.getDocumentMapper();
+
+        ParsedDocument doc = docMapper.parse(source(document)
+            .index(context.index().getName())
+            .id("_temp_id")
+            .type(documentType));
+
+        if (docMapperForType.getMapping() != null) {
+            doc.addDynamicMappingsUpdate(docMapperForType.getMapping());
+        }
+
+        Analyzer defaultAnalyzer = context.getAnalysisService().defaultIndexAnalyzer();
+        final IndexSearcher docSearcher;
+        if (doc.docs().size() > 1) {
+            assert docMapper.hasNestedObjects();
+            docSearcher = createMultiDocumentSearcher(docMapper, defaultAnalyzer, doc);
+        } else {
+            // TODO: we may want to bring to MemoryIndex thread local cache back...
+            // but I'm unsure about the real benefits.
+            MemoryIndex memoryIndex = new MemoryIndex(true);
+            indexDoc(docMapper, defaultAnalyzer, doc.rootDoc(), memoryIndex);
+            docSearcher = memoryIndex.createSearcher();
+            docSearcher.setQueryCache(null);
+        }
+
+        PercolatorQueryCache registry = context.getPercolatorQueryCache();
+        if (registry == null) {
+            throw new QueryShardException(context, "no percolator query registry");
+        }
+
+        Query percolateTypeQuery = new TermQuery(new Term(TypeFieldMapper.NAME, PercolatorFieldMapper.TYPE_NAME));
+        PercolatorQuery.Builder builder = new PercolatorQuery.Builder(
+                documentType, registry, document, docSearcher, percolateTypeQuery
+        );
+        Settings indexSettings = registry.getIndexSettings().getSettings();
+        if (indexSettings.getAsVersion(IndexMetaData.SETTING_VERSION_CREATED, null).onOrAfter(Version.V_5_0_0)) {
```

Comment:
the reason I did this is because otherwise checkStyle complains that the line is longer than 140 characters :)